# HACKER NEWS STATISTICS PROJECT

In this project, I am going to analyze the [Hacker News](https://news.ycombinator.com/) site. Basically, Hacker News is a website allow their user to posts the technology stories (known as "posts") and the other member can comment and vote, similar to reddit. Hacker News is extremely well-known in technology field, the top posts are usually attracted hundreds of thousands of visitors.

In this project I will compare two popular types of posts (`ask hn` and `show hn`) in Hacker News and find out which type witnesses more average comments per post. Besides, I also clarify the best specific time in a day for these two kind of posts to receive more attention from the other members.

__1. Reading and displaying data__

In [1]:
file = open("HN_posts_year_to_Sep_26_2016.csv", encoding='utf8')
hn_new=file.readlines()
hn=[]
for i in hn_new:
    i=i.replace("\n","")
    line= list(i.split(","))
    if line[-3] != '0':
        hn.append(line)
hn_header=hn[0]
hn=hn[1:]
len(hn)

80401

In [2]:
def explore_data(dataset, start=0, end=5, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print('\n') # adds a new (empty) line between rows
        print(row)        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
#This function help us to de=isplay the list of lists in a good-looking form,
#Easier to manage the data

In [3]:
print(hn_header)
print(explore_data(hn,0,3,True))

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


['12578975', 'Saving the Hassle of Shopping', 'https://blog.menswr.com/2016/09/07/whats-new-with-your-style-feed/', '1', '1', 'bdoux', '9/26/2016 3:13']


['12578908', 'Ask HN: What TLD do you use for local development?', '', '4', '7', 'Sevrene', '9/26/2016 2:53']


['12578822', 'Amazons Algorithms Dont Find You the Best Deals', 'https://www.technologyreview.com/s/602442/amazons-algorithms-dont-find-you-the-best-deals/', '1', '1', 'yarapavan', '9/26/2016 2:26']
Number of rows: 80401
Number of columns: 7
None


__2. Extracting `ask hn` and `show hn` posts__

In [4]:
ask_posts=[]
show_posts=[]
other_posts=[]
for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [5]:
len(ask_posts)

6382

In [6]:
total_row_in_different_lists=(len(show_posts)+len(ask_posts)+len(other_posts))
total_row=len(hn)
print(total_row==total_row_in_different_lists)

True


__3. Calculating the average comments per posts of two types__

In [7]:
total_ask_comments=0
for row in ask_posts:
    num_comments=float(row[-3])
    total_ask_comments+=num_comments
avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)

13.942337825133187


In [8]:
total_show_comments=0
for row in show_posts:
    num_comments=float(row[-3])
    total_show_comments+=num_comments
avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)

9.802930056710775


On average, the ask posts receive more comments compared to the number of the same criteria from show posts (13.74 > 9.81)

__4.1. Analysising data of `Ask Posts` by hour created__

In [9]:
time_and_comment_in_ask_hn=[]
for row in ask_posts:
    created_at=row[6]
    num_comments=int(row[-3])
    time_and_comment_in_ask_hn.append([created_at,num_comments])

In [10]:
print(time_and_comment_in_ask_hn[0:5])

[['9/26/2016 2:53', 7], ['9/26/2016 1:17', 3], ['9/25/2016 22:48', 3], ['9/25/2016 21:50', 2], ['9/25/2016 19:22', 22]]


We need to convert the strings in `'created_at'` column into 'datetime' datatype

In [11]:
import datetime as dt
counts_by_hour_ask_hn={}
comments_by_hour_ask_hn={}
for each in time_and_comment_in_ask_hn:
    time=each[0]
    comments_number=each[1]
    date_result=dt.datetime.strptime(time,"%m/%d/%Y %H:%M")
    hour_created=date_result.strftime("%H")
    if hour_created not in counts_by_hour_ask_hn:
        counts_by_hour_ask_hn[hour_created]=1
        comments_by_hour_ask_hn[hour_created]=comments_number
    else:
        counts_by_hour_ask_hn[hour_created]+=1
        comments_by_hour_ask_hn[hour_created]+=comments_number

In [12]:
print(counts_by_hour_ask_hn)

{'02': 212, '01': 199, '22': 268, '21': 368, '19': 397, '17': 366, '15': 435, '14': 350, '13': 296, '11': 234, '10': 208, '09': 164, '07': 148, '03': 198, '16': 379, '08': 180, '00': 216, '23': 255, '20': 362, '18': 419, '12': 251, '04': 165, '06': 161, '05': 151}


In [13]:
print(comments_by_hour_ask_hn)

{'02': 2922, '01': 1778, '22': 3202, '21': 3798, '19': 3852, '17': 5061, '15': 18073, '14': 4709, '13': 7028, '11': 2710, '10': 2869, '09': 1398, '07': 1543, '03': 1659, '16': 4089, '08': 2244, '00': 1805, '23': 2081, '20': 4292, '18': 4633, '12': 3893, '04': 2201, '06': 1500, '05': 1640}


Calculating average number of comments per hour in a day

In [14]:
avg_by_hour_ask_hn=[]
for each_row in comments_by_hour_ask_hn:
    if each_row in counts_by_hour_ask_hn:
        avg_by_hour_ask_hn.append([each_row, comments_by_hour_ask_hn[each_row]/counts_by_hour_ask_hn[each_row]])

In [15]:
avg_by_hour_ask_hn

[['02', 13.783018867924529],
 ['01', 8.93467336683417],
 ['22', 11.947761194029852],
 ['21', 10.320652173913043],
 ['19', 9.702770780856422],
 ['17', 13.827868852459016],
 ['15', 41.54712643678161],
 ['14', 13.454285714285714],
 ['13', 23.743243243243242],
 ['11', 11.581196581196581],
 ['10', 13.79326923076923],
 ['09', 8.524390243902438],
 ['07', 10.425675675675675],
 ['03', 8.378787878787879],
 ['16', 10.78891820580475],
 ['08', 12.466666666666667],
 ['00', 8.356481481481481],
 ['23', 8.16078431372549],
 ['20', 11.856353591160222],
 ['18', 11.05727923627685],
 ['12', 15.50996015936255],
 ['04', 13.33939393939394],
 ['06', 9.316770186335404],
 ['05', 10.860927152317881]]

Sorting the list above base on the second value of of each list (the average number of comments per hour)

In [16]:
def func(lists):
    return lists[1]
sorted_swap_ask_hn=sorted(avg_by_hour_ask_hn,key=func,reverse=True)

In [17]:
sorted_swap_ask_hn

[['15', 41.54712643678161],
 ['13', 23.743243243243242],
 ['12', 15.50996015936255],
 ['17', 13.827868852459016],
 ['10', 13.79326923076923],
 ['02', 13.783018867924529],
 ['14', 13.454285714285714],
 ['04', 13.33939393939394],
 ['08', 12.466666666666667],
 ['22', 11.947761194029852],
 ['20', 11.856353591160222],
 ['11', 11.581196581196581],
 ['18', 11.05727923627685],
 ['05', 10.860927152317881],
 ['16', 10.78891820580475],
 ['07', 10.425675675675675],
 ['21', 10.320652173913043],
 ['19', 9.702770780856422],
 ['06', 9.316770186335404],
 ['01', 8.93467336683417],
 ['09', 8.524390243902438],
 ['03', 8.378787878787879],
 ['00', 8.356481481481481],
 ['23', 8.16078431372549]]

In [18]:
top_5_highest_chance_time_ask_hn=[]
for i in sorted_swap_ask_hn[0:5]:
    time=dt.datetime.strptime(i[0],"%H")
    time_string=time.strftime("%H:00")
    platform="{}: {:.2f} average comments per post"
    string=platform.format(time_string,i[1])
    top_5_highest_chance_time_ask_hn.append(string)

In [19]:
print("Top 5 Hours for 'Ask HN' Comments")
explore_data(top_5_highest_chance_time_ask_hn)

Top 5 Hours for 'Ask HN' Comments


15:00: 41.55 average comments per post


13:00: 23.74 average comments per post


12:00: 15.51 average comments per post


17:00: 13.83 average comments per post


10:00: 13.79 average comments per post


__4.2. Analysising data of `Show Posts` by hour created__

Now we just need to execute the same process as we did for `Ask Posts`

In [20]:
time_and_comment_in_show_hn=[]
for row in show_posts:
    created_at=row[6]
    num_comments=int(row[-3])
    time_and_comment_in_show_hn.append([created_at,num_comments])

In [21]:
counts_by_hour_show_hn={}
comments_by_hour_show_hn={}
for each in time_and_comment_in_show_hn:
    times=each[0]
    comments_numbers=each[1]
    date_result=dt.datetime.strptime(times,"%m/%d/%Y %H:%M")
    hour_created=date_result.strftime("%H")
    if hour_created not in counts_by_hour_show_hn:
        counts_by_hour_show_hn[hour_created]=1
        comments_by_hour_show_hn[hour_created]=comments_numbers
    else:
        counts_by_hour_show_hn[hour_created]+=1
        comments_by_hour_show_hn[hour_created]+=comments_numbers

ValueError: time data 'askstories' does not match format '%m/%d/%Y %H:%M'

Oops! It's seem like there is a error cell in show posts. I will find that flaw.

In [22]:
for i in time_and_comment_in_show_hn:
    i_1=i[0]
    if i_1 == "askstories":
        print(time_and_comment_in_show_hn.index(i))

294


In [23]:
time_and_comment_in_show_hn[294]

['askstories', 2]

I can totally remove this error line from the `time_and_comment_in_show_hn` to make sure the code run well, however, in case if there are another error data, I will edit the code above a little bit

In [24]:
import datetime as dt

counts_by_hour_show_hn={}
comments_by_hour_show_hn={}
for each in time_and_comment_in_show_hn:
    try:
        times=each[0]
        comments_numbers=each[1]
        date_result=dt.datetime.strptime(times,"%m/%d/%Y %H:%M")
        hour_created=date_result.strftime("%H")
        if hour_created not in counts_by_hour_show_hn:
            counts_by_hour_show_hn[hour_created]=1
            comments_by_hour_show_hn[hour_created]=comments_numbers
        else:
            counts_by_hour_show_hn[hour_created]+=1
            comments_by_hour_show_hn[hour_created]+=comments_numbers
    except ValueError:
        pass

In [25]:
print(counts_by_hour_show_hn)

{'20': 203, '19': 222, '16': 313, '10': 124, '06': 74, '03': 81, '18': 263, '17': 312, '15': 330, '11': 187, '14': 294, '13': 287, '09': 132, '01': 109, '22': 164, '12': 247, '08': 134, '04': 74, '00': 123, '21': 182, '05': 65, '23': 119, '07': 105, '02': 85}


In [26]:
print(comments_by_hour_show_hn)

{'20': 1790, '19': 2323, '16': 2936, '10': 980, '06': 730, '03': 742, '18': 2842, '17': 2660, '15': 2951, '11': 2162, '14': 3223, '13': 2934, '09': 1057, '01': 786, '22': 1310, '12': 3010, '08': 1517, '04': 788, '00': 1122, '21': 1653, '05': 521, '23': 1208, '07': 1453, '02': 781}


Now, it's clear!

In [27]:
avg_by_hour_show_hn=[]
for i in comments_by_hour_show_hn:
    if i in counts_by_hour_show_hn:
        avg_by_hour_show_hn.append([i, comments_by_hour_show_hn[i]/counts_by_hour_show_hn[i]])

In [28]:
avg_by_hour_show_hn

[['20', 8.817733990147783],
 ['19', 10.463963963963964],
 ['16', 9.380191693290735],
 ['10', 7.903225806451613],
 ['06', 9.864864864864865],
 ['03', 9.160493827160494],
 ['18', 10.806083650190114],
 ['17', 8.525641025641026],
 ['15', 8.942424242424243],
 ['11', 11.561497326203208],
 ['14', 10.962585034013605],
 ['13', 10.222996515679442],
 ['09', 8.007575757575758],
 ['01', 7.2110091743119265],
 ['22', 7.987804878048781],
 ['12', 12.186234817813766],
 ['08', 11.32089552238806],
 ['04', 10.64864864864865],
 ['00', 9.121951219512194],
 ['21', 9.082417582417582],
 ['05', 8.015384615384615],
 ['23', 10.15126050420168],
 ['07', 13.838095238095239],
 ['02', 9.188235294117646]]

In [29]:
def func(lists):
    return lists[1]
sorted_swap_show_hn=sorted(avg_by_hour_show_hn,key=func,reverse=True)

In [30]:
sorted_swap_show_hn

[['07', 13.838095238095239],
 ['12', 12.186234817813766],
 ['11', 11.561497326203208],
 ['08', 11.32089552238806],
 ['14', 10.962585034013605],
 ['18', 10.806083650190114],
 ['04', 10.64864864864865],
 ['19', 10.463963963963964],
 ['13', 10.222996515679442],
 ['23', 10.15126050420168],
 ['06', 9.864864864864865],
 ['16', 9.380191693290735],
 ['02', 9.188235294117646],
 ['03', 9.160493827160494],
 ['00', 9.121951219512194],
 ['21', 9.082417582417582],
 ['15', 8.942424242424243],
 ['20', 8.817733990147783],
 ['17', 8.525641025641026],
 ['05', 8.015384615384615],
 ['09', 8.007575757575758],
 ['22', 7.987804878048781],
 ['10', 7.903225806451613],
 ['01', 7.2110091743119265]]

In [31]:
top_5_highest_chance_time_show_hn=[]
for i in sorted_swap_show_hn[0:5]:
    time=dt.datetime.strptime(i[0],"%H")
    time_string=time.strftime("%H:00")
    platform="{}: {:.2f} average comments per post"
    string=platform.format(time_string,i[1])
    top_5_highest_chance_time_show_hn.append(string)

In [32]:
print("Top 5 Hours for 'show HN' Comments")
explore_data(top_5_highest_chance_time_show_hn)

Top 5 Hours for 'show HN' Comments


07:00: 13.84 average comments per post


12:00: 12.19 average comments per post


11:00: 11.56 average comments per post


08:00: 11.32 average comments per post


14:00: 10.96 average comments per post


As you can see from the result in __4.1__ and __4.2__, `ask posts` receives more considerations than `show posts` does. The highest average number of comments in timeframe (at 07:00) of `show posts` (13.84) is only approximately equal to the fourth place (13.83) of the highest figure for `ask posts`, which is at 17:00. Besides, the timeframe holding the highest figure in `ask post` is totally dominant compared to the second position of the same type of post, specifically, about 75% higher of the the second place.

# Conclusion

I am living in Finland (GMT +2), and timezone of the documentation was Eastern Time in the US (GMT -5). So which mean the best time for me to create an ask post was 22:00, for a show post was 14:00